File: samples/language-impl-comparison/rust/markdown_parser.rs
// Markdown風軽量マークアップパーサー - Rust実装
//
// Unicode処理の注意点：
// - Rustの標準String/strは内部的にUTF-8だが、charはUnicodeスカラー値（コードポイント）
// - 書記素クラスター（Grapheme）の扱いにはunicode-segmentationクレートが必要
// - Remlの3層モデル（Byte/Char/Grapheme）と比較すると、Rustは明示的な区別が必要

use std::fmt;

/// Markdown AST のブロック要素。
#[derive(Debug, Clone, PartialEq)]
pub enum Block {
    Heading { level: usize, inline: Vec<Inline> },
    Paragraph { inline: Vec<Inline> },
    UnorderedList { items: Vec<Vec<Inline>> },
    OrderedList { items: Vec<Vec<Inline>> },
    CodeBlock { lang: Option<String>, code: String },
    HorizontalRule,
}

/// Markdown AST のインライン要素。
#[derive(Debug, Clone, PartialEq)]
pub enum Inline {
    Text(String),
    Strong(Vec<Inline>),
    Emphasis(Vec<Inline>),
    Code(String),
    Link { text: Vec<Inline>, url: String },
    LineBreak,
}

pub type Document = Vec<Block>;

#[derive(Debug)]
pub enum ParseError {
    UnexpectedEof,
    InvalidHeading(String),
    InvalidCodeBlock(String),
    Other(String),
}

impl fmt::Display for ParseError {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        match self {
            ParseError::UnexpectedEof => write!(f, "予期しないファイル終端"),
            ParseError::InvalidHeading(msg) => write!(f, "見出しエラー: {}", msg),
            ParseError::InvalidCodeBlock(msg) => write!(f, "コードブロックエラー: {}", msg),
            ParseError::Other(msg) => write!(f, "{}", msg),
        }
    }
}

/// パーサー状態。
///
/// Unicode処理の課題：
/// - positionはバイト位置（Rustのstrスライスの制約）
/// - char単位の処理には.chars()イテレータが必要
/// - Grapheme単位の処理には外部クレート（unicode-segmentation）が必要
struct Parser<'a> {
    input: &'a str,
    position: usize,  // バイト位置（Rustの制約）
}

impl<'a> Parser<'a> {
    fn new(input: &'a str) -> Self {
        Parser { input, position: 0 }
    }

    /// 現在位置の1文字（char = Unicodeスカラー値）を取得。
    ///
    /// 注意：Rustのcharはコードポイントだが、Graphemeではない。
    /// 「🇯🇵」のような国旗絵文字は2つのcharとして扱われる。
    fn peek(&self) -> Option<char> {
        self.input[self.position..].chars().next()
    }

    /// 1文字（char）を消費して進める。
    ///
    /// 課題：UTF-8の可変長エンコーディングのため、
    /// len_utf8()でバイト長を計算する必要がある。
    fn advance_char(&mut self) {
        if let Some(ch) = self.peek() {
            self.position += ch.len_utf8();
        }
    }

    /// 固定文字列をマッチ。
    fn match_string(&mut self, target: &str) -> bool {
        if self.input[self.position..].starts_with(target) {
            self.position += target.len();
            true
        } else {
            false
        }
    }

    /// 水平空白をスキップ。
    fn skip_hspace(&mut self) {
        while let Some(ch) = self.peek() {
            if ch == ' ' || ch == '\t' {
                self.advance_char();
            } else {
                break;
            }
        }
    }

    /// 行末まで読む。
    fn read_until_eol(&mut self) -> String {
        let start = self.position;
        while let Some(ch) = self.peek() {
            if ch == '\n' {
                break;
            }
            self.advance_char();
        }
        self.input[start..self.position].to_string()
    }

    /// 改行を消費。
    fn consume_newline(&mut self) -> bool {
        if self.peek() == Some('\n') {
            self.advance_char();
            true
        } else {
            false
        }
    }

    /// EOFチェック。
    fn is_eof(&self) -> bool {
        self.position >= self.input.len()
    }

    /// 見出し行のパース（`# Heading` 形式）。
    fn parse_heading(&mut self) -> Result<Block, ParseError> {
        self.skip_hspace();

        // `#` の連続をカウント
        let mut level = 0;
        while self.peek() == Some('#') {
            level += 1;
            self.advance_char();
        }

        if level == 0 || level > 6 {
            return Err(ParseError::InvalidHeading(
                "見出しレベルは1-6の範囲内である必要があります".to_string(),
            ));
        }

        self.skip_hspace();
        let text = self.read_until_eol();
        self.consume_newline();

        Ok(Block::Heading {
            level,
            inline: vec![Inline::Text(text.trim().to_string())],
        })
    }

    /// 水平線のパース（`---`, `***`, `___`）。
    fn parse_horizontal_rule(&mut self) -> Result<Block, ParseError> {
        self.skip_hspace();
        let text = self.read_until_eol();
        self.consume_newline();

        let trimmed = text.trim();
        let is_rule = (trimmed.chars().all(|c| c == '-') && trimmed.len() >= 3)
            || (trimmed.chars().all(|c| c == '*') && trimmed.len() >= 3)
            || (trimmed.chars().all(|c| c == '_') && trimmed.len() >= 3);

        if is_rule {
            Ok(Block::HorizontalRule)
        } else {
            Err(ParseError::Other("水平線として認識できません".to_string()))
        }
    }

    /// コードブロックのパース（```言語名）。
    fn parse_code_block(&mut self) -> Result<Block, ParseError> {
        if !self.match_string("```") {
            return Err(ParseError::InvalidCodeBlock(
                "コードブロック開始が見つかりません".to_string(),
            ));
        }

        let lang_line = self.read_until_eol();
        self.consume_newline();

        let lang = if lang_line.trim().is_empty() {
            None
        } else {
            Some(lang_line.trim().to_string())
        };

        // コードブロック内容を ```閉じまで読む
        let mut code_lines = Vec::new();
        loop {
            if self.is_eof() {
                break;
            }
            if self.match_string("```") {
                break;
            }
            let line = self.read_until_eol();
            code_lines.push(line);
            self.consume_newline();
        }

        let code = code_lines.join("\n");
        self.consume_newline();

        Ok(Block::CodeBlock { lang, code })
    }

    /// リスト項目のパース（簡易版：`-` または `*`）。
    fn parse_unordered_list(&mut self) -> Result<Block, ParseError> {
        let mut items = Vec::new();

        loop {
            self.skip_hspace();
            match self.peek() {
                Some('-') | Some('*') => {
                    self.advance_char();
                    self.skip_hspace();
                    let text = self.read_until_eol();
                    self.consume_newline();
                    items.push(vec![Inline::Text(text.trim().to_string())]);
                }
                _ => break,
            }
        }

        if items.is_empty() {
            Err(ParseError::Other("リスト項目が見つかりません".to_string()))
        } else {
            Ok(Block::UnorderedList { items })
        }
    }

    /// 段落のパース（簡易版：空行まで）。
    fn parse_paragraph(&mut self) -> Result<Block, ParseError> {
        let mut lines = Vec::new();

        loop {
            if self.is_eof() {
                break;
            }
            if self.peek() == Some('\n') {
                self.advance_char();
                if self.peek() == Some('\n') {
                    break;
                }
                lines.push("".to_string());
            } else {
                let line = self.read_until_eol();
                lines.push(line);
                self.consume_newline();
            }
        }

        let text = lines.join(" ");
        Ok(Block::Paragraph {
            inline: vec![Inline::Text(text.trim().to_string())],
        })
    }

    /// ブロック要素のパース（優先順位付き試行）。
    fn parse_block(&mut self) -> Result<Block, ParseError> {
        // 空行スキップ
        while self.peek() == Some('\n') {
            self.advance_char();
        }

        if self.is_eof() {
            return Err(ParseError::UnexpectedEof);
        }

        self.skip_hspace();

        match self.peek() {
            Some('#') => self.parse_heading(),
            Some('`') if self.input[self.position..].starts_with("```") => self.parse_code_block(),
            Some('-') | Some('*') | Some('_') => {
                // 水平線かリストか判定
                let saved_pos = self.position;
                match self.parse_horizontal_rule() {
                    Ok(block) => Ok(block),
                    Err(_) => {
                        self.position = saved_pos;
                        self.parse_unordered_list()
                    }
                }
            }
            _ => self.parse_paragraph(),
        }
    }
}

/// ドキュメント全体のパース。
pub fn parse(input: &str) -> Result<Document, ParseError> {
    let mut parser = Parser::new(input);
    let mut blocks = Vec::new();

    loop {
        match parser.parse_block() {
            Ok(block) => blocks.push(block),
            Err(ParseError::UnexpectedEof) => break,
            Err(e) => return Err(e),
        }
    }

    Ok(blocks)
}

/// 簡易的なレンダリング（検証用）。
pub fn render_to_string(doc: &Document) -> String {
    fn render_inline(inline: &[Inline]) -> String {
        inline
            .iter()
            .map(|i| match i {
                Inline::Text(s) => s.clone(),
                Inline::Strong(inner) => format!("**{}**", render_inline(inner)),
                Inline::Emphasis(inner) => format!("*{}*", render_inline(inner)),
                Inline::Code(s) => format!("`{}`", s),
                Inline::Link { text, url } => format!("[{}]({})", render_inline(text), url),
                Inline::LineBreak => "\n".to_string(),
            })
            .collect()
    }

    doc.iter()
        .map(|block| match block {
            Block::Heading { level, inline } => {
                format!("{} {}\n\n", "#".repeat(*level), render_inline(inline))
            }
            Block::Paragraph { inline } => format!("{}\n\n", render_inline(inline)),
            Block::UnorderedList { items } => {
                let mut result = String::new();
                for item in items {
                    result.push_str(&format!("- {}\n", render_inline(item)));
                }
                result.push('\n');
                result
            }
            Block::OrderedList { items } => {
                let mut result = String::new();
                for (i, item) in items.iter().enumerate() {
                    result.push_str(&format!("{}. {}\n", i + 1, render_inline(item)));
                }
                result.push('\n');
                result
            }
            Block::CodeBlock { lang, code } => {
                let lang_str = lang.as_deref().unwrap_or("");
                format!("```{}\n{}\n```\n\n", lang_str, code)
            }
            Block::HorizontalRule => "---\n\n".to_string(),
        })
        .collect()
}

// Rustでのグラフェムクラスター処理の課題を示すヘルパー関数
// （unicode-segmentationクレートが必要だが、ここではコメントで示す）

/// Unicode 3層モデル比較：
///
/// Remlでは Byte/Char/Grapheme が型レベルで区別されるが、
/// Rustでは明示的なクレート使用が必要：
///
/// ```rust
/// // use unicode_segmentation::UnicodeSegmentation;
///
/// // pub fn count_graphemes(text: &str) -> usize {
/// //     text.graphemes(true).count()
/// // }
///
/// // pub fn count_codepoints(text: &str) -> usize {
/// //     text.chars().count()
/// // }
///
/// // pub fn byte_length(text: &str) -> usize {
/// //     text.len()
/// // }
/// ```
///
/// この明示性の欠如が、絵文字や結合文字の扱いでバグを生む可能性がある。

File: examples/language-impl-comparison/scala3/markdown_parser.scala
// Markdown風軽量マークアップパーサー - Scala 3実装
//
// Unicode処理の注意点：
// - ScalaのStringはJavaのString（UTF-16ベース）
// - length はUTF-16コードユニット数を返す（サロゲートペアに注意）
// - codePointCount でUnicodeコードポイント数を取得可能
// - Grapheme（書記素クラスター）処理には外部ライブラリ（icu4j等）が必要
// - Remlの3層モデル（Byte/Char/Grapheme）と比較すると、Scala 3も明示的な区別が必要

package markdownParser

// Markdown AST のインライン要素
enum Inline:
  case Text(text: String)
  case Strong(inner: List[Inline])
  case Emphasis(inner: List[Inline])
  case Code(code: String)
  case Link(text: List[Inline], url: String)
  case LineBreak

// Markdown AST のブロック要素
enum Block:
  case Heading(level: Int, inline: List[Inline])
  case Paragraph(inline: List[Inline])
  case UnorderedList(items: List[List[Inline]])
  case OrderedList(items: List[List[Inline]])
  case CodeBlock(lang: Option[String], code: String)
  case HorizontalRule

type Document = List[Block]

// パーサー状態
case class ParseState(input: String, position: Int)

// パーサー例外
case class ParseError(message: String) extends Exception(message)

// 現在位置の1文字を取得
def peekChar(state: ParseState): Option[Char] =
  if state.position >= state.input.length then None
  else Some(state.input(state.position))

// 1文字を消費して進める
def advanceChar(state: ParseState): ParseState =
  state.copy(position = state.position + 1)

// 固定文字列をマッチ
def matchString(state: ParseState, target: String): Option[ParseState] =
  val remaining = state.input.substring(state.position)
  if remaining.startsWith(target) then
    Some(state.copy(position = state.position + target.length))
  else
    None

// 水平空白をスキップ
def skipHSpace(state: ParseState): ParseState =
  peekChar(state) match
    case Some(' ') | Some('\t') => skipHSpace(advanceChar(state))
    case _ => state

// 空行をスキップ
def skipBlankLines(state: ParseState): ParseState =
  peekChar(state) match
    case Some('\n') => skipBlankLines(advanceChar(state))
    case _ => state

// 行末まで読む
def readUntilEol(state: ParseState): (String, ParseState) =
  def loop(pos: Int, acc: StringBuilder): (String, ParseState) =
    if pos >= state.input.length then
      (acc.toString, state.copy(position = pos))
    else if state.input(pos) == '\n' then
      (acc.toString, state.copy(position = pos))
    else
      loop(pos + 1, acc.append(state.input(pos)))

  loop(state.position, new StringBuilder)

// 改行を消費
def consumeNewline(state: ParseState): ParseState =
  peekChar(state) match
    case Some('\n') => advanceChar(state)
    case _ => state

// EOFチェック
def isEof(state: ParseState): Boolean =
  state.position >= state.input.length

// 見出し行のパース（`# Heading` 形式）
def parseHeading(state: ParseState): (Block, ParseState) =
  val state1 = skipHSpace(state)

  // `#` の連続をカウント
  def countHashes(s: ParseState, n: Int): (Int, ParseState) =
    peekChar(s) match
      case Some('#') => countHashes(advanceChar(s), n + 1)
      case _ => (n, s)

  val (level, state2) = countHashes(state1, 0)

  if level == 0 || level > 6 then
    throw ParseError("見出しレベルは1-6の範囲内である必要があります")
  else
    val state3 = skipHSpace(state2)
    val (text, state4) = readUntilEol(state3)
    val state5 = consumeNewline(state4)
    val inline = List(Inline.Text(text.trim))
    (Block.Heading(level, inline), state5)

// 水平線のパース（`---`, `***`, `___`）
def parseHorizontalRule(state: ParseState): (Block, ParseState) =
  val state1 = skipHSpace(state)
  val (text, state2) = readUntilEol(state1)
  val state3 = consumeNewline(state2)

  val trimmed = text.trim
  val isRule =
    (trimmed.forall(_ == '-') && trimmed.length >= 3) ||
    (trimmed.forall(_ == '*') && trimmed.length >= 3) ||
    (trimmed.forall(_ == '_') && trimmed.length >= 3)

  if !isRule then
    throw ParseError("水平線として認識できません")
  else
    (Block.HorizontalRule, state3)

// コードブロックのパース（```言語名）
def parseCodeBlock(state: ParseState): (Block, ParseState) =
  matchString(state, "```") match
    case None => throw ParseError("コードブロック開始が見つかりません")
    case Some(state1) =>
      val (langLine, state2) = readUntilEol(state1)
      val state3 = consumeNewline(state2)

      val lang =
        val trimmed = langLine.trim
        if trimmed.isEmpty then None else Some(trimmed)

      // コードブロック内容を ```閉じまで読む
      def readCodeLines(s: ParseState, lines: List[String]): (List[String], ParseState) =
        matchString(s, "```") match
          case Some(endState) => (lines.reverse, endState)
          case None =>
            if isEof(s) then
              (lines.reverse, s)
            else
              val (line, s2) = readUntilEol(s)
              val s3 = consumeNewline(s2)
              readCodeLines(s3, line :: lines)

      val (codeLines, state4) = readCodeLines(state3, Nil)
      val state5 = consumeNewline(state4)

      val code = codeLines.mkString("\n")
      (Block.CodeBlock(lang, code), state5)

// リスト項目のパース（簡易版：`-` または `*`）
def parseUnorderedList(state: ParseState): (Block, ParseState) =
  def parseItems(s: ParseState, items: List[List[Inline]]): (List[List[Inline]], ParseState) =
    val s1 = skipHSpace(s)
    peekChar(s1) match
      case Some('-') | Some('*') =>
        val s2 = advanceChar(s1)
        val s3 = skipHSpace(s2)
        val (text, s4) = readUntilEol(s3)
        val s5 = consumeNewline(s4)
        val inline = List(Inline.Text(text.trim))
        parseItems(s5, items :+ inline)
      case _ => (items, s)

  val (items, stateEnd) = parseItems(state, Nil)

  if items.isEmpty then
    throw ParseError("リスト項目が見つかりません")
  else
    (Block.UnorderedList(items), stateEnd)

// 段落のパース（簡易版：空行まで）
def parseParagraph(state: ParseState): (Block, ParseState) =
  def readLines(s: ParseState, lines: List[String]): (List[String], ParseState) =
    if isEof(s) then
      (lines.reverse, s)
    else
      peekChar(s) match
        case Some('\n') =>
          val s1 = advanceChar(s)
          peekChar(s1) match
            case Some('\n') => (lines.reverse, s1) // 空行で段落終了
            case _ => readLines(s1, "" :: lines)
        case Some(_) =>
          val (line, s2) = readUntilEol(s)
          val s3 = consumeNewline(s2)
          readLines(s3, line :: lines)
        case None => (lines.reverse, s)

  val (lines, stateEnd) = readLines(state, Nil)
  val text = lines.mkString(" ").trim
  val inline = List(Inline.Text(text))
  (Block.Paragraph(inline), stateEnd)

// ブロック要素のパース（優先順位付き試行）
def parseBlock(state: ParseState): (Block, ParseState) =
  val state1 = skipBlankLines(state)

  if isEof(state1) then
    throw ParseError("EOF")
  else
    val state2 = skipHSpace(state1)

    peekChar(state2) match
      case Some('#') => parseHeading(state2)
      case Some('`') =>
        matchString(state2, "```") match
          case Some(_) => parseCodeBlock(state2)
          case None => parseParagraph(state2)
      case Some('-') | Some('*') | Some('_') =>
        try
          parseHorizontalRule(state2)
        catch
          case _: ParseError => parseUnorderedList(state2)
      case Some(_) => parseParagraph(state2)
      case None => throw ParseError("EOF")

// ドキュメント全体のパース
def parseDocument(state: ParseState, blocks: List[Block]): Document =
  try
    val (block, newState) = parseBlock(state)
    parseDocument(newState, blocks :+ block)
  catch
    case ParseError("EOF") => blocks

// パブリックAPI：文字列からドキュメントをパース
def parse(input: String): Document =
  val initialState = ParseState(input, 0)
  parseDocument(initialState, Nil)

// 簡易的なレンダリング（検証用）
def renderInline(inlines: List[Inline]): String =
  inlines.map {
    case Inline.Text(s) => s
    case Inline.Strong(inner) => s"**${renderInline(inner)}**"
    case Inline.Emphasis(inner) => s"*${renderInline(inner)}*"
    case Inline.Code(s) => s"`$s`"
    case Inline.Link(text, url) => s"[${renderInline(text)}]($url)"
    case Inline.LineBreak => "\n"
  }.mkString

def renderBlock(block: Block): String =
  block match
    case Block.Heading(level, inline) =>
      val prefix = "#" * level
      s"$prefix ${renderInline(inline)}\n\n"
    case Block.Paragraph(inline) =>
      s"${renderInline(inline)}\n\n"
    case Block.UnorderedList(items) =>
      val itemsStr = items.map(item => s"- ${renderInline(item)}\n").mkString
      s"$itemsStr\n"
    case Block.OrderedList(items) =>
      val itemsStr = items.zipWithIndex.map { case (item, i) =>
        s"${i + 1}. ${renderInline(item)}\n"
      }.mkString
      s"$itemsStr\n"
    case Block.CodeBlock(lang, code) =>
      val langStr = lang.getOrElse("")
      s"```$langStr\n$code\n```\n\n"
    case Block.HorizontalRule =>
      "---\n\n"

def renderToString(doc: Document): String =
  doc.map(renderBlock).mkString

// Unicode 3層モデル比較：
//
// ScalaのStringはJavaのString（UTF-16ベース）で：
// - length はUTF-16コードユニット数を返す
// - サロゲートペア（例：絵文字「😀」）は2カウントされる
// - codePointCount でUnicodeコードポイント数を取得可能
// - Grapheme処理には icu4j 等の外部ライブラリが必要
//
// 例：
// val str = "🇯🇵"  // 国旗絵文字（2つのコードポイント、1つのgrapheme）
// str.length  // => 4 (UTF-16コードユニット数)
// str.codePointCount(0, str.length)  // => 2 (コードポイント数)
//
// Remlの3層モデル（Byte/Char/Grapheme）と比較すると、
// Scala 3も明示的な区別が必要で、標準APIだけでは絵文字や結合文字の扱いが複雑。
//
// Reml との比較メモ:
// 1. Scala 3: enum で代数的データ型を簡潔に表現
//    Reml: 型定義で直接 `type Inline = Text(string) | Strong(...) | ...` と記述
//    - 両言語とも現代的な構文で、パターンマッチが強力
// 2. Scala 3: immutableデータ構造とcopyメソッドで状態を管理
//    Reml: 関数型 + 手続き型のハイブリッドアプローチ
// 3. Scala 3: Either型やTry型でエラーハンドリング、例外も利用可能
//    Reml: Result型を標準で提供し、? 演算子で簡潔に記述
// 4. 両言語とも型推論が強力で、型安全性を確保